File: snippets/cpp/mst/prim.cpp
ll prim(vector<vpii> &g, int s = 0) {
    int n = sza(g); vi vis(n); ll cost = 0;
    priority_queue<pii, vpii, greater<pii>> pq; pq.push({0, s});
    while (!pq.empty()) {
        auto [w, u] = pq.top(); pq.pop();
        if (vis[u]) continue;
        vis[u] = 1; cost += w;
        for (auto [v, wt] : g[u])
            if (!vis[v]) pq.push({wt, v});
    }
    return cost;
}

File: snippets/rust/mst/prim.rs
// Prim's MST using a min-heap. `g[u]` is a list of (v, w) outgoing edges.
fn prim(g: &Vec<Vec<(usize, i64)>>, s: usize) -> i64 {
    use std::cmp::Reverse;
    use std::collections::BinaryHeap;
    let n = g.len();
    let mut vis = vec![false; n];
    let mut pq: BinaryHeap<Reverse<(i64, usize)>> = BinaryHeap::new();
    pq.push(Reverse((0, s)));
    let mut cost = 0i64;
    while let Some(Reverse((w, u))) = pq.pop() {
        if vis[u] { continue; }
        vis[u] = true;
        cost += w;
        for &(v, ww) in &g[u] {
            if !vis[v] { pq.push(Reverse((ww, v))); }
        }
    }
    cost
}
